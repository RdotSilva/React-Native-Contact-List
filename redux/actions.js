// Action Types
export const UPDATE_USER = 'UPDATE_USER';
export const UPDATE_CONTACT = 'UPDATE_CONTACT';
export const LOG_IN_SENT = 'LOG_IN_SENT';
export const LOG_IN_FULFILLED = 'LOG_IN_FULFILLED';
export const LOG_IN_REJECTED = 'LOG_IN_REJECTED';

//Action Creators
export const updateUser = update => ({
	type: UPDATE_USER,
	payload: update
});

export const addContact = newContact => ({
	type: UPDATE_CONTACT,
	payload: newContact
});

// Async Action Creator
const logInUser = (username, password) => dispatch => {
	dispatch({ type: 'LOG_IN_SENT' });
	login(username, password)
		.then(() => {
			dispatch({ type: 'LOG_IN_SUCCESS' });
		})
		.catch(err => {
			dispatch({ type: 'LOG_IN_REJECTED' });
		});
};
