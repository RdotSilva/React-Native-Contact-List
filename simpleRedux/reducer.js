const reducer = (state, update) => ({
	...state,
	...update
});
