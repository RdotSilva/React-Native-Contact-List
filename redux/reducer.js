import { combineReducers } from 'redux';
import {
	UPDATE_USER,
	UPDATE_CONTACT,
	LOG_IN_SENT,
	LOG_IN_FULFILLED,
	LOG_IN_REJECTED,
	CHANGE_FIRST_CONTACT
} from './actions';

import contacts from '../contacts';

const contactReducer = (state = contacts, action) => {
	if (action.type === UPDATE_CONTACT) {
		return [...state, action.payload];
	}
	if (action.type === CHANGE_FIRST_CONTACT) {
		const [firstContact, ...rest] = state;
		if (!firstContact) return state;
		// [{name: 'jordan', phone: '1234567890'}]
		// firstContact, rest = []
		const newContact = { ...firstContact, name: 'Overwrite Name' };
		return [newContact, ...rest];
	}
	return state;
};

const userReducer = (state = {}, action) => {
	switch (action.type) {
		case UPDATE_USER:
			return {
				...state,
				...action.payload
			};
		case UPDATE_CONTACT:
			return {
				...state,
				prevContact: action.payload
			};
		case LOG_IN_FULFILLED:
			return {
				...state,
				token: action.payload
			};
		case LOG_IN_REJECTED:
			return {
				...state,
				loginErr: action.payload
			};
		default:
			return state;
	}
};

// Main reducer that combines all reducers
const reducer = combineReducers({
	user: userReducer,
	contacts: contactReducer
});

export default reducer;
