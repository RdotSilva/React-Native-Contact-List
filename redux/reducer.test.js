import reducer from './reducer';
import * as actions from './actions';

describe('contact reducer', () => {
	it('successfully adds new user', () => {});
});
