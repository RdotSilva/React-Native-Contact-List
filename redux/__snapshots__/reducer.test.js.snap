// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contact reducer successfully adds new user 1`] = `
Object {
  "contacts": Array [
    Object {
      "name": "test user",
      "phone": "1234567890",
    },
  ],
  "user": Object {
    "prevContact": Object {
      "name": "test user",
      "phone": "1234567890",
    },
  },
}
`;
